crates/turborepo-lib/src/turbo_json/parser.rs

```diff
@@ -0,0 +1,556 @@
+use thiserror::Error;
+
+use crate::{
+    cli::OutputLogsMode,
+    config::ConfigurationOptions,
+    run::task_id::TaskName,
+    turbo_json::{Pipeline, RawTaskDefinition, RawTurboJson, SpacesJson, Spanned},
+};
+
+#[derive(Debug, Error, Diagnostic)]
+pub enum Error {
+    #[error("failed to parse turbo.json")]
+    Parse {
+        diagnostics: Vec<biome_diagnostics::Error>,
+    },
+}
+
+fn print_diagnostics(diagnostics: &[biome_diagnostics::Error], color: bool) {
+    let color_mode = if color {
+        ColorMode::Enabled
+    } else {
+        ColorMode::Disabled
+    };
+    let mut console = EnvConsole::new(color_mode);
+    for diagnostic in diagnostics {
+        console.error(markup!({ PrintDiagnostic::simple(diagnostic) }));
+    }
+}
+
+impl<T: Deserializable> Deserializable for Spanned<T> {
+    fn deserialize(
+        value: &impl DeserializableValue,
+        name: &str,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        let range = value.range();
+        let value = T::deserialize(value, name, diagnostics)?;
+        Some(Spanned {
+            value,
+            range: Some(range.into()),
+            path: None,
+            text: None,
+        })
+    }
+}
+
+impl Deserializable for OutputLogsMode {
+    fn deserialize(
+        value: &impl DeserializableValue,
+        name: &str,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        match String::deserialize(value, name, diagnostics)?.as_str() {
+            "full" => Some(OutputLogsMode::Full),
+            "none" => Some(OutputLogsMode::None),
+            "hash-only" => Some(OutputLogsMode::HashOnly),
+            "new-only" => Some(OutputLogsMode::NewOnly),
+            "errors-only" => Some(OutputLogsMode::ErrorsOnly),
+            unknown_variant => {
+                const ALLOWED_VARIANTS: &[&str] =
+                    &["full", "none", "hash-only", "new-only", "errors-only"];
+                diagnostics.push(DeserializationDiagnostic::new_unknown_value(
+                    unknown_variant,
+                    value.range(),
+                    ALLOWED_VARIANTS,
+                ));
+                None
+            }
+        }
+    }
+}
+
+impl Deserializable for TaskName<'static> {
+    fn deserialize(
+        value: &impl DeserializableValue,
+        name: &str,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        let task_id = String::deserialize(value, name, diagnostics)?;
+
+        Some(Self::from(task_id))
+    }
+}
+
+impl Deserializable for RawTaskDefinition {
+    fn deserialize(
+        value: &impl DeserializableValue,
+        name: &str,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        value.deserialize(RawTaskDefinitionVisitor, name, diagnostics)
+    }
+}
+
+struct RawTaskDefinitionVisitor;
+
+impl DeserializationVisitor for RawTaskDefinitionVisitor {
```

Comment:
Is there a way we could leverage a macro to reduce to amount of handwritten code for these implementations? This is a lot of handwritten code where it's really easy to mess up in some minor way e.g. forget `.with_range` on a `Spanned` value or forgetting to add a new key to `ALLOWED_KEYS`.